src/backend/distributed/utils/multi_partitioning_utils.c

```diff
@@ -0,0 +1,362 @@
+#include "distributed/citus_ruleutils.h"
+#include <distributed/multi_partitioning_utils.h>
+#include "lib/stringinfo.h"
+#include "nodes/pg_list.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+#include "utils/syscache.h"
+
+
+#if (PG_VERSION_NUM >= 100000)
+static char * PartitionBound(Oid partitionId);
+#endif
+
+
+/*
+ * Returns true if the given relation is a partitioned table.
+ */
+bool
+PartitionedTable(Oid relationId)
+{
+	Relation rel = heap_open(relationId, AccessShareLock);
+	bool partitionedTable = false;
+
+#if (PG_VERSION_NUM >= 100000)
+	if (rel->rd_rel->relkind == RELKIND_PARTITIONED_TABLE)
+	{
+		partitionedTable = true;
+	}
+#endif
+
+	/* keep the lock */
+	heap_close(rel, NoLock);
+
+	return partitionedTable;
+}
+
+
+/*
+ * Returns true if the given relation is a partition.
+ */
+bool
+PartitionTable(Oid relationId)
+{
+	Relation rel = heap_open(relationId, AccessShareLock);
+	bool partitionTable = false;
+
+#if (PG_VERSION_NUM >= 100000)
+	partitionTable = rel->rd_rel->relispartition;
+#endif
+
+	/* keep the lock */
+	heap_close(rel, NoLock);
+
+	return partitionTable;
+}
+
+
+/*
+ * IsChildTable returns true if the table is inherited. Note that
+ * partition tables inherites by default. However, this function
+ * returns false if the given table is a partition.
+ */
+bool
+IsChildTable(Oid relationId)
+{
+	Relation pgInherits = NULL;
+	SysScanDesc scan = NULL;
+	ScanKeyData key[1];
+	HeapTuple inheritsTuple = NULL;
+	bool tableInherits = false;
+
+	pgInherits = heap_open(InheritsRelationId, AccessShareLock);
+
+	ScanKeyInit(&key[0], Anum_pg_inherits_inhrelid,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(relationId));
+
+	scan = systable_beginscan(pgInherits, InvalidOid, false,
+							  NULL, 1, key);
+
+	while ((inheritsTuple = systable_getnext(scan)) != NULL)
+	{
+		Oid inheritedRelationId =
+			((Form_pg_inherits) GETSTRUCT(inheritsTuple))->inhrelid;
+
+		if (relationId == inheritedRelationId)
+		{
+			tableInherits = true;
+			break;
+		}
+	}
+
+	systable_endscan(scan);
+	heap_close(pgInherits, AccessShareLock);
+
+	if (tableInherits && PartitionTable(relationId))
+	{
+		tableInherits = false;
+	}
+
+	return tableInherits;
+}
+
+
+/*
+ * IsParentTable returns true if the table is inherited. Note that
+ * partitioned tables inherited by default. However, this function
+ * returns false if the given table is a partitioned table.
+ */
+bool
+IsParentTable(Oid relationId)
+{
+	Relation pgInherits = NULL;
+	SysScanDesc scan = NULL;
+	ScanKeyData key[1];
+	HeapTuple inheritsTuple = NULL;
+	bool tableInherited = false;
+
+	pgInherits = heap_open(InheritsRelationId, AccessShareLock);
+
+	ScanKeyInit(&key[0], Anum_pg_inherits_inhparent,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(relationId));
+
+	scan = systable_beginscan(pgInherits, InheritsParentIndexId, true,
+							  NULL, 1, key);
+
+	while ((inheritsTuple = systable_getnext(scan)) != NULL)
```

Comment:
Doing `if (systable_getnext(scan) != NULL) tableInherited = true)` should be sufficient here.